internal/serverinstall/docker.go

```diff
@@ -240,6 +240,201 @@ func (i *DockerInstaller) Install(
 	}, nil
 }
 
+// Upgrade is a method of DockerInstaller and implements the Installer interface to
+// upgrade a waypoint-server as a Docker container
+func (i *DockerInstaller) Upgrade(
+	ctx context.Context, opts *InstallOpts, serverCfg serverconfig.Client) (
+	*InstallResults, error,
+) {
+	ui := opts.UI
+
+	sg := ui.StepGroup()
+	defer sg.Wait()
+
+	s := sg.Add("Initializing Docker client...")
+	defer func() { s.Abort() }()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, err
+	}
+	cli.NegotiateAPIVersion(ctx)
+
+	s.Update("Checking for an existing Waypoint server installation...")
+	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "label",
+			Value: "waypoint-type=server",
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	grpcPort := defaultGrpcPort
+	httpPort := defaultHttpPort
+
+	var (
+		clicfg   clicontext.Config
+		addr     pb.ServerConfig_AdvertiseAddr
+		httpAddr string
+	)
+
+	clicfg.Server = serverconfig.Client{
+		Address:       "localhost:" + grpcPort,
+		Tls:           true,
+		TlsSkipVerify: true,
+	}
+
+	addr.Addr = serverName + ":" + grpcPort
+	addr.Tls = true
+	addr.TlsSkipVerify = true
+
+	httpAddr = "localhost:" + httpPort
+
+	if len(containers) < 0 {
+		s.Update("No waypoint server detected. Nothing to upgrade.")
+		s.Status(terminal.StatusWarn)
+		s.Done()
+		return nil, fmt.Errorf("No waypoint server container detected")
+	}
+
+	// Assume waypoint-server is the first container with the waypoint-type label
+	waypointServerContainer := containers[0]
+
+	s.Update("Checking for Docker image: %s", i.config.serverImage)
+
+	imageRef, err := reference.ParseNormalizedNamed(i.config.serverImage)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing Docker image: %s", err)
+	}
+
+	imageList, err := cli.ImageList(ctx, types.ImageListOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{
+			Key:   "reference",
+			Value: reference.FamiliarString(imageRef),
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(imageList) == 0 || i.config.serverImage == defaultServerImage {
+		s = sg.Add("Pulling image: %s", i.config.serverImage)
+
+		resp, err := cli.ImagePull(ctx, reference.FamiliarString(imageRef), types.ImagePullOptions{})
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Close()
+
+		stdout, _, err := ui.OutputWriters()
+		if err != nil {
+			return nil, err
+		}
+
+		var termFd uintptr
+		if f, ok := stdout.(*os.File); ok {
+			termFd = f.Fd()
+		}
+
+		err = jsonmessage.DisplayJSONMessagesStream(resp, s.TermOutput(), termFd, true, nil)
+		if err != nil {
+			return nil, fmt.Errorf("unable to stream pull logs to the terminal: %s", err)
+		}
+		s.Done()
+		s = sg.Add("")
+	}
+
+	s.Update("Upgrading Waypoint server image from %q to %q",
+		waypointServerContainer.Image, i.config.serverImage)
```

Comment:
nit: readability -- put each item on its own line since we're already breaking the line. i can't make a suggestion bc I can't gofmt it afterwards 😄 , but i mean:
```
(
  "Upgrading Waypoint server image from %q to %q",
  waypointServerContainer.Image,
  i.config.serverImage,
)
```